tech-specs/src/explore/design/dkg.md

```diff
@@ -1 +1,154 @@
+\\( W\\) (and fractional weights rounded appropriately to acheive integral
+values). After this procedure, each validator has a corresponding number of
+weight shares. 
+
+Lastly, the DKG protocol requires a canonical ordering of validators with
+shares must be agreed upon. This will be done via the consensus layer.
+The current block proposer will apportion out weight shares as well as the
+aforementioned ordering (which induces a natural partition of the weight 
+shares) to each validator that made an announcement. This is included in the
+block. Once this block is on the blockchain, the next step of the DKG can proceed.
+
+## Dealers
+In a single DKG instance, a subset of validators holding at least
+\\( \frac{2}{3} \\) of the weight shares must each act as dealers. It
+is the responsibility of a dealer to post a signed message to the blockchain
+which includes the session id (\\( \tau \\)) as well the data of the
+PVSS (we do not concern ourselves with the details here). The creation of this
+message is handled by the Ferveo library.
+
+This means that the signed message is a transaction that must be gossiped and
+eventually included on the blockchain. For this particular transaction, there
+is a special verification that must be done, otherwise the transaction should
+not be included in any proposed block. This verification is provided by the
+Ferveo library.
+
+## Aggregation
+
+At some point, enough validators (meaning their combined weight  shares
+exceeds \\( \frac{2}{3}\\) of the total) have their PVSS on the chain 
+for the session with id \\( \tau \\). At this point, the next block proposer can
+aggregate the messages into a single instance using Ferveo. This is then 
+included in the next block. Once this DKG is on the blockchain, it can be 
+used for encryption of transactions until the next DKG instance is run.
+
+When deciding to aggregate the PVSS instances into a single one, we shall 
+consider the instances ranked in terms of the weights of their dealers (in 
+descending order). The topmost dealers constituting enough weight are then 
+chosen and aggregated in their specified order.
+
+We note that enough weight may be gossiped to the current proposer and thus
+they can put the PVSS transactions and aggregation all in the same proposed 
+block.
+
+## Notes on pipelining
+
+Note that in the decription above, at least two blocks must be submitted to the blockchain in order to
+produce the next public encryption key. The first block determines the 
+partition of weight shares of the validators and the last aggregates the 
+PVSS instances from the dealers leading to the final key.
+
+In practice, it may take even longer as the protocol waits for enough PVSS 
+instances to get onto the blockchain. Depending on the chosen refresh rate
+of the keys, another instance of the DKG protocol may need to be started 
+before the current one finishes. This is possible so long as all data
```

Comment:
I don't think we want DKG instances to overlap, though? (At least, I sincerely hope the DKG doesn't take a full day to run)

I think the problem is that the public key for epoch i must be created by running the DKG during epoch i-1, which presents the following issues:

1. The validator set during epoch i-1 might be very different than epoch i, so in the worst case scenario maybe the old validator set censors all the DKG messages sent by the new validator set. So for safety there must be a fallback for the DKG to keep going into the new epoch (if necessary).
2. The staking weights for epoch i must be finalized before the DKG for epoch i begins

Reply:
I included this section before it was clear how often we would generate keys. But now that once per epoch should be the target, this is much less critical. However, I agree we need a fallback in the case of radical validator changes. I suppose we just use the old key a bit longer until a new one can be created?

Reply:
Actually, there is no guarantee that we can decrypt on old key either. I'm not sure what we should do in this case. The only two options I see is 
1. Accept unencrypted transactions until a new key is made
2. Accept no transactions until a new key is made.

Reply:
I think this is related to the broader question of what happens to old transactions in the mempool that don't make it on chain in time before the key expires.

Unfortunately I think there are not many options other than not accepting new transactions - hopefully only a few blocks have to be empty before the new DKG finishes, and it only happens because 2/3 of validators in the previous epoch chose to behave maliciously (in which case they can do a lot more bad stuff)